### PR TITLE
Provide a fast CI option for draft status

### PR DESCRIPTION
**Pull request by h-mayorquin**
This goes further than #1078.

The idea is the following. A two-tier CI depending on PR status (draft vs ready for review)

When the PR is on draft we only test on a minimal version of Python so simple errors can surface quickly.

Once the PR is ready for a review then we run the full python matrix to ensure that no error escapes the CI.

@pauladkisson thoughts? 
I actually thought about going further and not run dev or live tests on draft but I first wanted to illustrate the idea.

https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/evaluate-expressions-in-workflows-and-actions#example

### CODE REVIEW

File: .github/workflows/deploy-tests.yml [hunk 1]
```diff
@@ -37,14 +39,25 @@ jobs:
...
       CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
-
+    with:
+      python-versions: >  # Ternary operator,  condition && value_if_true || value_if_false
+        ${{github.event.pull_request.draft == true
+           && '["3.9"]'
+           || '["3.9", "3.10", "3.11", "3.12"]'
+        }}
 
   run-live-service-tests:
```
Comment (pauladkisson) [inline .github/workflows/deploy-tests.yml:49]: honestly, do we need live-service-tests or dev-tests while a PR is in draft?

In fact, we might not need them at all in the PR tests.  They could just be dailies...

**Review by pauladkisson — CHANGES_REQUESTED** [review]
I like this idea overall, just added a few suggestions.

File: .github/workflows/deploy-tests.yml [hunk 2]
```diff
@@ -37,14 +39,25 @@ jobs:
       AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
       S3_GIN_BUCKET: ${{ secrets.S3_GIN_BUCKET }}
       CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
-
+    with:
+      python-versions: >  # Ternary operator,  condition && value_if_true || value_if_false
+        ${{github.event.pull_request.draft == true
+           && '["3.9"]'
```
[the remark this answers is not in the document]
Comment (h-mayorquin) [reply .github/workflows/deploy-tests.yml:45]: My thinking here is that most of us do not use the oldest version of Python.  This prevents us from getting errors because we use things not supported in Python 3.9 when we develop.

File: .github/workflows/testing.yml [hunk 3]
```diff
@@ -10,7 +17,6 @@ on:
         required: true
       CODECOV_TOKEN:
         required: true
-  workflow_dispatch:
```
[the remark this answers is not in the document]
Comment (h-mayorquin) [reply .github/workflows/testing.yml:13]: I should not, probably a mistake. Thanks for the catch.

File: .github/workflows/testing.yml [hunk 4]
```diff
@@ -19,7 +25,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: ["3.9", "3.10", "3.11", "3.12"]
+        python-version: ${{ fromJson(inputs.python-versions) }}
         os: [ubuntu-latest, macos-13, windows-latest]
```
[the remark this answers is not in the document]
Comment (h-mayorquin) [reply .github/workflows/testing.yml:29]: I like this, and I think that in neuroconv it makes sense because we do use all the oses as developers (I use Linux, you and Ben use Mac, Cody uses Windows I think that Alessandra uses Windows as well, not sure what Szjona uses).

That said, Windows and Linux are sufficiently different and I would like to catch those errors early. 

A compromise would be to do Linux and Windows latest and then once the PR is ready add arm and intel for Mac. What do you think?

File: .github/workflows/deploy-tests.yml (hunk 1 above)
Comment (h-mayorquin) [reply .github/workflows/deploy-tests.yml:49]: I think this makes sense. Let me think a bit.

File: .github/workflows/testing.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (pauladkisson) [reply .github/workflows/testing.yml:30]: > That said, Windows and Linux are sufficiently different and I would like to catch those errors early.

I thought the whole point of paring down the testing matrix while a PR is on draft is bc there are likely to be actual bugs in the code (not related to external factors like os, python version, etc.) that should be resolved first, while the substantive development happens. Then, when the PR is ready for review, we can do a full test matrix to make sure that it passes under all relevant conditions (python version, os, etc).

In other words, if there are intrinsic bugs in the code, they can be caught by one OS and one python version, which saves us a whole ton of runners.

File: .github/workflows/deploy-tests.yml (hunk 1 above)
Comment (pauladkisson) [reply .github/workflows/deploy-tests.yml:49]: sounds good, separate PR is fine.

File: .github/workflows/testing.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (h-mayorquin) [reply .github/workflows/testing.yml:30]: I was not thinking exactly like that. Maybe your approach makes more sense but this is how I was thinking:

There must be a Goldilocks zone of testing that detects most of the early errors or you code quickly and without a lot of CI overhead. Some use cases to make it concrete:
1) Environment issues. Your local environment assumes something that is not true. 
2) Your code fixes your problem but breaks the code somewhere else.
3) CI issues that can't be tested locally.
4) The user does not have gin-data or wants to see if the gin-data is propagated to the CI.

Once you have fixed this and you feel more confident about the PR then the full tests in ready-to-review catch the long-tail of the errors if there is anything.

From my experience, there are some shallow errors that a Linux coder can make that don't work on Windows (hard coding paths posix style with `/` being a typical one) and that's why I wanted to have Windows.

But this is not that strong of an argument to me. If you think that Linux only is better we can start with only Linux and then if many simple window errors appear we can add windows in the future. What do you think?

[the remark this answers is not in the document]
Comment (pauladkisson) [reply .github/workflows/testing.yml:30]: > There must be a Goldilocks zone of testing that detects most of the early errors or you code quickly and without a lot of CI overhead.

This framework makes sense, I just think that the best bang for our buck is with 1 OS. The majority of cases you are describing (1-4) should be caught regardless of OS, and Windows + Linux is twice as many tests as Linux-only.  Plus, I really don't see any concrete downside to fixing these os-specific bugs (like hard-coding posix-style paths) later in the development cycle.

[the remark this answers is not in the document]
Comment (h-mayorquin) [reply .github/workflows/testing.yml:30]: Those are good points, @pauladkisson. Let's move forward in that direction.

[the remark this answers is not in the document]
Comment (h-mayorquin) [reply .github/workflows/testing.yml]: OK, I did this in the last commit. Also, now that https://github.com/catalystneuro/neuroconv/pull/1061 is merged, I re-added mac with intel chips as discussed in https://github.com/catalystneuro/neuroconv/pull/1080/files.

If this works I will refactor the repeated lines using env variables and then this should be ready to go.

**Review by pauladkisson — APPROVED** [review]
lgtm!
